server/src/test/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshotsTest.java

```diff
@@ -0,0 +1,109 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package org.elasticsearch.index.snapshots.blobstore;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.elasticsearch.repositories.blobstore.BlobStoreRepository.INDEX_SHARD_SNAPSHOTS_FORMAT;
+
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.lucene.util.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+import org.elasticsearch.index.store.StoreFileMetadata;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Test;
+
+import com.sun.management.ThreadMXBean;
+
+public class BlobStoreIndexShardSnapshotsTest extends ESTestCase {
+
+
+    @Test
+    public void test_bwc_streaming() throws Exception {
+        BlobStoreIndexShardSnapshots blobStoreIndexShardSnapshots = prepareData();
+        BytesStreamOutput out = new BytesStreamOutput();
+        blobStoreIndexShardSnapshots.writeTo(out);
+
+        StreamInput in = out.bytes().streamInput();
+        BlobStoreIndexShardSnapshots fromStream = BlobStoreIndexShardSnapshots.fromStream(in);
+
+        //TODO: implement equals or compare per-field, also, try different versions
+        assertThat(blobStoreIndexShardSnapshots).isEqualTo(fromStream);
+    }
+
+    @Test
+    public void test_read_from_stream_allocations() throws Exception {
+        BytesReference bytesReference = INDEX_SHARD_SNAPSHOTS_FORMAT.serialize(prepareData(), "dummyBlobName", true);
+
+        ThreadMXBean threadMXBean = ManagementFactory.getPlatformMXBean(ThreadMXBean.class);
+        long threadId = Thread.currentThread().threadId();
+        long allocatedBytesBegin = threadMXBean.getThreadAllocatedBytes(threadId);
+
+        INDEX_SHARD_SNAPSHOTS_FORMAT.deserialize(
+            "dummyBlobname",
+            writableRegistry(),
+            xContentRegistry(),
+            bytesReference
+        );
+
+        long allocatedBytesAfter = threadMXBean.getThreadAllocatedBytes(threadId);
+        long allocatedBytes = allocatedBytesAfter - allocatedBytesBegin;
+        // Actual value fluctuates around 9_877_672.
+        // Used to fluctuate around 29_228_160 on 5.10.2 when instances were not re-used.
+        assertThat(allocatedBytes).isLessThanOrEqualTo(10_000_000L);
```

Comment:
3x difference, so no `getThreadAllocatedBytes` fluctuation. 

Could make it 10x but didn't want to create too many files in test, 10K is already enough